src/main/java/de/tum/in/www1/artemis/config/MetricsBean.java

```diff
@@ -113,13 +129,145 @@ private void registerWebsocketMetrics() {
                 .description("Number of users connected to this Artemis instance").register(meterRegistry);
     }
 
+    private void registerExerciseAndExamMetrics() {
+        int[] ranges = { 5, 15, 30, 45, 60, 120 };
+        for (int range : ranges) {
+            registerExerciseMetrics(range);
+            registerExamMetrics(range);
+        }
+    }
+
+    private void registerExerciseMetrics(int range) {
```

Comment:
This method will retrieve many duplicates from the database. For example every exercise that is within the next five minutes is also within the next 10 minutes.
Additionally you access the same courses multiple times. Wouldn't it be better to retrieve the courses and exercises once and then calculate the statistics?

Reply:
MicroMeter (the library that we use to expose the metrics) has two different concepts that we could use here.

We are currently using [Gauges](https://micrometer.io/docs/concepts#_gauges) here: They are defined once per `name` and `tags` (within this register method) and a supplier function is given that MicroMeter will execute whenever an update is needed (usually when the metrics are retrieved by e.g. the monitoring system).
The supplier function returns a `value` that will be exposed.
Each Gauge is independent of each other and thus, as you stated, duplicates will be retrieved.



The other approach is using a [MultiGauge](https://micrometer.io/docs/concepts#_multi_gauge): They are defined once per `name`, but tags are not added during the registration (tags added during the registration are added to all metrics the `MultiGauge` provides, but can not be used to differentiate the different metrics since they will be identical).
The `MultiGauge` can the be used to supply different metrics, consisting of `tags` and `values`.
We could therefor e.g. load all exercises within the next `120` minutes from the database once and then do a separation into exercise type and the 'smaller' ranges (5, 15, 30, 45, 60 minutes) based on this dataset since the range `120` also contains the other ranges as subsets.
This would decrease the database load since fewer queries have to be executed (each query itself is larger however, since we can not just let the database do the counting but have to return the whole exercise to then do the filtering in the application code).


The downside of `MultiGauges` compared to `Gauges` are that they will not be updated automatically by MicroMeter, instead we have to take care of updating the values on our own (we would define a method that updates the registered metrics and call this method whenever we want to update the exposed metrics).
In a deployment with a single instance of Artemis, we could trigger this update whenever a change to an exercise is made (creation, deletion, update of an exercise).
In our distributed setup this is not possible since each instance calculates the metrics independently of each other (and e.g. Kubernetes currently also expects it like this) and instances are not informed whenever another instance updates an exercise (at least not in a way we can listen to in a simple way).

This requires us to periodically schedule the update on each instance.
If we, e.g., schedule an update every 60 seconds and also let the monitoring system query the metrics every 60 seconds, this would introduce an additional delay (which should be 30 seconds, on average) until the monitoring system retrieves the updated metrics.

@krusche What do you think would be the approprioate approach here?
`Gauge` (current approach):
- More database queries (that return less data each)
- Instant updates

`MultiGauge`:
- fewer database queries (that return more data each)
- Updates are delayed (on average `updateInterval / 2`)